File: common/src/main/java/org/apache/gravitino/dto/requests/ViewUpdateRequest.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
package org.apache.gravitino.dto.requests;

import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
import com.fasterxml.jackson.annotation.JsonProperty;
import com.fasterxml.jackson.annotation.JsonSubTypes;
import com.fasterxml.jackson.annotation.JsonTypeInfo;
import com.google.common.base.Preconditions;
import java.util.Arrays;
import java.util.HashSet;
import java.util.Set;
import javax.annotation.Nullable;
import lombok.EqualsAndHashCode;
import lombok.Getter;
import lombok.ToString;
import org.apache.commons.lang3.StringUtils;
import org.apache.gravitino.dto.rel.ColumnDTO;
import org.apache.gravitino.dto.rel.RepresentationDTO;
import org.apache.gravitino.dto.rel.SQLRepresentationDTO;
import org.apache.gravitino.dto.util.DTOConverters;
import org.apache.gravitino.rel.ViewChange;
import org.apache.gravitino.rest.RESTRequest;

/** Represents a request to update a view. */
@JsonIgnoreProperties(ignoreUnknown = true)
@JsonTypeInfo(use = JsonTypeInfo.Id.NAME, include = JsonTypeInfo.As.PROPERTY)
@JsonSubTypes({
  @JsonSubTypes.Type(value = ViewUpdateRequest.RenameViewRequest.class, name = "rename"),
  @JsonSubTypes.Type(value = ViewUpdateRequest.SetViewPropertyRequest.class, name = "setProperty"),
  @JsonSubTypes.Type(
      value = ViewUpdateRequest.RemoveViewPropertyRequest.class,
      name = "removeProperty"),
  @JsonSubTypes.Type(value = ViewUpdateRequest.ReplaceViewRequest.class, name = "replaceView")
})
public interface ViewUpdateRequest extends RESTRequest {

  /**
   * The view change represented by this request.
   *
   * @return An instance of {@link ViewChange}.
   */
  ViewChange viewChange();

  /** Represents a request to rename a view. */
  @EqualsAndHashCode
  @ToString
  @Getter
  class RenameViewRequest implements ViewUpdateRequest {

    @JsonProperty("newName")
    private final String newName;

    /**
     * Constructor for RenameViewRequest.
     *
     * @param newName The new name of the view.
     */
    public RenameViewRequest(String newName) {
      this.newName = newName;
    }

    /** Default constructor for Jackson deserialization. */
    public RenameViewRequest() {
      this(null);
    }

    @Override
    public void validate() throws IllegalArgumentException {
      Preconditions.checkArgument(
          StringUtils.isNotBlank(newName), "\"newName\" field is required and cannot be empty");
    }

    @Override
    public ViewChange viewChange() {
      return ViewChange.rename(newName);
    }
  }

  /** Represents a request to set a property of a view. */
  @EqualsAndHashCode
  @ToString
  @Getter
  class SetViewPropertyRequest implements ViewUpdateRequest {

    @JsonProperty("property")
    private final String property;

    @JsonProperty("value")
    private final String value;

    /**
     * Constructor for SetViewPropertyRequest.
     *
     * @param property The property to set.
     * @param value The value of the property.
     */
    public SetViewPropertyRequest(String property, String value) {
      this.property = property;
      this.value = value;
    }

    /** Default constructor for Jackson deserialization. */
    public SetViewPropertyRequest() {
      this(null, null);
    }

    @Override
    public void validate() throws IllegalArgumentException {
      Preconditions.checkArgument(
          StringUtils.isNotBlank(property), "\"property\" field is required and cannot be empty");
      Preconditions.checkArgument(value != null, "\"value\" field is required and cannot be null");
    }

    @Override
    public ViewChange viewChange() {
      return ViewChange.setProperty(property, value);
    }
  }

  /** Represents a request to remove a property of a view. */
  @EqualsAndHashCode
  @ToString
  @Getter
  class RemoveViewPropertyRequest implements ViewUpdateRequest {

    @JsonProperty("property")
    private final String property;

    /**
     * Constructor for RemoveViewPropertyRequest.
     *
     * @param property The property to remove.
     */
    public RemoveViewPropertyRequest(String property) {
      this.property = property;
    }

    /** Default constructor for Jackson deserialization. */
    public RemoveViewPropertyRequest() {
      this(null);
    }

    @Override
    public void validate() throws IllegalArgumentException {
      Preconditions.checkArgument(
          StringUtils.isNotBlank(property), "\"property\" field is required and cannot be empty");
    }

    @Override
    public ViewChange viewChange() {
      return ViewChange.removeProperty(property);
    }
  }

  /**
   * Represents a request to atomically replace the body (columns, representations, default catalog,
   * default schema and comment) of a view. View name and properties are not affected.
   */
  @EqualsAndHashCode
  @ToString
  @Getter
  class ReplaceViewRequest implements ViewUpdateRequest {

    @JsonProperty("columns")
    private final ColumnDTO[] columns;

    @JsonProperty("representations")
    private final RepresentationDTO[] representations;

    @JsonProperty("defaultCatalog")
    @Nullable
    private final String defaultCatalog;

    @JsonProperty("defaultSchema")
    @Nullable
    private final String defaultSchema;

    @JsonProperty("comment")
    @Nullable
    private final String comment;

    /**
     * Constructor for ReplaceViewRequest.
     *
     * @param columns The new output columns of the view.
     * @param representations The new representations of the view.
     * @param defaultCatalog The new default catalog, or {@code null} to unset it.
     * @param defaultSchema The new default schema, or {@code null} to unset it.
     * @param comment The new comment, or {@code null} to unset it.
     */
    public ReplaceViewRequest(
        ColumnDTO[] columns,
        RepresentationDTO[] representations,
        @Nullable String defaultCatalog,
        @Nullable String defaultSchema,
        @Nullable String comment) {
      this.columns = columns;
      this.representations = representations;
      this.defaultCatalog = defaultCatalog;
      this.defaultSchema = defaultSchema;
      this.comment = comment;
    }

    /** Default constructor for Jackson deserialization. */
    public ReplaceViewRequest() {
      this(null, null, null, null, null);
    }

    @Override
    public void validate() throws IllegalArgumentException {
      Preconditions.checkArgument(
          representations != null && representations.length > 0,
          "\"representations\" field is required and cannot be empty");
      Arrays.stream(representations)
          .forEach(
              rep -> {
                Preconditions.checkArgument(rep != null, "representation must not be null");
                rep.validate();
              });

      Set<String> seenDialects = new HashSet<>();
      Arrays.stream(representations)
          .filter(rep -> rep instanceof SQLRepresentationDTO)
          .map(rep -> ((SQLRepresentationDTO) rep).dialect())
          .forEach(
              dialect ->
                  Preconditions.checkArgument(
                      seenDialects.add(dialect),
                      "Duplicate SQL representation dialect: %s",
                      dialect));
    }

    @Override
    public ViewChange viewChange() {
      return ViewChange.replaceView(
          columns == null ? new ColumnDTO[0] : columns,
          DTOConverters.fromDTOs(representations),
          defaultCatalog,
          defaultSchema,
          comment);
    }
  }
}

File: clients/client-python/gravitino/dto/requests/view_update_request.py
# Licensed to the Apache Software Foundation (ASF) under one
# or more contributor license agreements.  See the NOTICE file
# distributed with this work for additional information
# regarding copyright ownership.  The ASF licenses this file
# to you under the Apache License, Version 2.0 (the
# "License"); you may not use this file except in compliance
# with the License.  You may obtain a copy of the License at
#
#   http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing,
# software distributed under the License is distributed on an
# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
# KIND, either express or implied.  See the License for the
# specific language governing permissions and limitations
# under the License.

from abc import ABC, abstractmethod
from dataclasses import dataclass, field
from typing import Optional

from dataclasses_json import config, dataclass_json

from gravitino.api.rel.view_change import ViewChange
from gravitino.dto.rel.column_dto import ColumnDTO
from gravitino.dto.rel.json_serdes.representation_serdes import RepresentationSerdes
from gravitino.dto.rel.representation_dto import RepresentationDTO
from gravitino.dto.requests.view_create_request import ViewCreateRequest
from gravitino.rest.rest_message import RESTRequest
from gravitino.utils.precondition import Precondition


@dataclass_json
@dataclass
class ViewUpdateRequestBase(RESTRequest, ABC):
    """Base class for all view update requests."""

    _type: str = field(init=False, metadata=config(field_name="@type"))

    @abstractmethod
    def view_change(self) -> ViewChange:
        """Convert to view change operation."""


class ViewUpdateRequest:
    """Namespace for all view update request types."""

    @dataclass_json
    @dataclass
    class RenameViewRequest(ViewUpdateRequestBase):
        """Update request to rename a view."""

        _new_name: str = field(metadata=config(field_name="newName"))

        def __post_init__(self):
            self._type = "rename"

        def validate(self):
            Precondition.check_string_not_empty(
                self._new_name, '"newName" field is required and cannot be empty'
            )

        def view_change(self) -> ViewChange:
            return ViewChange.rename(self._new_name)

    @dataclass_json
    @dataclass
    class SetViewPropertyRequest(ViewUpdateRequestBase):
        """Update request to set a view property."""

        _property: str = field(metadata=config(field_name="property"))
        _value: str = field(metadata=config(field_name="value"))

        def __post_init__(self):
            self._type = "setProperty"

        def validate(self):
            Precondition.check_string_not_empty(
                self._property, '"property" field is required and cannot be empty'
            )
            Precondition.check_argument(
                self._value is not None, '"value" field is required and cannot be null'
            )

        def view_change(self) -> ViewChange:
            return ViewChange.set_property(self._property, self._value)

    @dataclass_json
    @dataclass
    class RemoveViewPropertyRequest(ViewUpdateRequestBase):
        """Update request to remove a view property."""

        _property: str = field(metadata=config(field_name="property"))

        def __post_init__(self):
            self._type = "removeProperty"

        def validate(self):
            Precondition.check_string_not_empty(
                self._property, '"property" field is required and cannot be empty'
            )

        def view_change(self) -> ViewChange:
            return ViewChange.remove_property(self._property)

    @dataclass_json
    @dataclass
    class ReplaceViewRequest(ViewUpdateRequestBase):
        """Update request to replace the view body."""

        _columns: Optional[list[ColumnDTO]] = field(
            default=None, metadata=config(field_name="columns")
        )
        _representations: list[RepresentationDTO] = field(
            default=None,
            metadata=config(
                field_name="representations",
                encoder=lambda items: [
                    RepresentationSerdes.serialize(item) for item in items
                ],
            ),
        )
        _default_catalog: Optional[str] = field(
            default=None, metadata=config(field_name="defaultCatalog")
        )
        _default_schema: Optional[str] = field(
            default=None, metadata=config(field_name="defaultSchema")
        )
        _comment: Optional[str] = field(
            default=None, metadata=config(field_name="comment")
        )

        def __post_init__(self):
            self._type = "replaceView"

        def validate(self):
            Precondition.check_argument(
                self._representations is not None and len(self._representations) > 0,
                '"representations" field is required and cannot be empty',
            )
            for representation in self._representations:
                Precondition.check_argument(
                    representation is not None, "representation must not be null"
                )
                representation.validate()
            if self._columns:
                for column in self._columns:
                    Precondition.check_argument(
                        column is not None, "column must not be null"
                    )
                    column.validate()
            ViewCreateRequest.validate_no_duplicate_dialects(self._representations)

        def view_change(self) -> ViewChange:
            return ViewChange.replace_view(
                self._columns or [],
                self._representations,
                self._default_catalog,
                self._default_schema,
                self._comment,
            )
